src/base/bittorrent/session.cpp

```diff
@@ -1887,8 +1889,58 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
 
         p = magnetUri.addTorrentParams();
     }
-    else if (torrentInfo.isValid()) {
-        if (!params.restored) {
+    else {
+        if (!torrentInfo.isValid()) {
+            // We can have an invalid torrentInfo when there isn't a matching
+            // .torrent file to the .fastresume we loaded. Possibly from a
+            // failed upgrade.
+            return false;
+        }
+
+        hash = torrentInfo.hash();
+    }
+
+    // We should not add the torrent if it is already
+    // processed or is pending to add to session
+    if (m_addingTorrents.contains(hash) || m_loadedMetadata.contains(hash))
+        return false;
+
+    TorrentHandle *const torrent = m_torrents.value(hash);
+    if (torrent) {  // a duplicate torrent is added
+        if (torrent->isPrivate() || (!fromMagnetUri && torrentInfo.isPrivate()))
+            return false;
+
+        // merge trackers and web seeds
+        torrent->addTrackers(fromMagnetUri ? magnetUri.trackers() : torrentInfo.trackers());
+        torrent->addUrlSeeds(fromMagnetUri ? magnetUri.urlSeeds() : torrentInfo.urlSeeds());
+        return true;
+    }
+
+    // Detect if fastresume contains downloaded metadata
+    // In the case of magent metadata isn't fetched yet, we still need
+    // to setup the default parameters.
+    bool isMetadataDownloaded = false;
```

Comment:
I wouldn't use "downloaded" suffix here. It confuses me a bit. Additionally we really concern not about metadata but about fastresume data. I would use "isFastresumeAvailable".

Reply:
>Additionally we really concern not about metadata but about fastresume data. I would use "isFastresumeAvailable".

There is fastresume data and it is valid, what we don't have is the torrent metadata. I'll rename it to `isMetadataAvailable`.

Reply:
I answered you in other comment.